File: 03_variable_selection_regularization/code/regularization_comparison.py
import numpy as np
import pandas as pd
from sklearn.linear_model import Ridge, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt
import seaborn as sns

def demonstrate_regularization_comparison():
    """Demonstrate ridge vs lasso regularization comparison"""
    
    # Generate synthetic data
    np.random.seed(42)
    n, p = 100, 20
    X = np.random.randn(n, p)
    true_beta = np.zeros(p)
    true_beta[:5] = [2, -1.5, 1, -0.8, 0.6]  # Only first 5 coefficients are non-zero
    y = X @ true_beta + 0.5 * np.random.randn(n)

    # Split data
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

    # Standardize data
    scaler_X = StandardScaler()
    scaler_y = StandardScaler()

    X_train_scaled = scaler_X.fit_transform(X_train)
    X_test_scaled = scaler_X.transform(X_test)
    y_train_scaled = scaler_y.fit_transform(y_train.reshape(-1, 1)).ravel()
    y_test_scaled = scaler_y.transform(y_test.reshape(-1, 1)).ravel()

    # Fit models
    lambda_values = np.logspace(-3, 3, 50)

    # Ridge regression
    ridge_scores = []
    ridge_coefs = []

    for alpha in lambda_values:
        ridge = Ridge(alpha=alpha)
        scores = cross_val_score(ridge, X_train_scaled, y_train_scaled, cv=5, scoring='neg_mean_squared_error')
        ridge_scores.append(-scores.mean())
        
        ridge.fit(X_train_scaled, y_train_scaled)
        ridge_coefs.append(ridge.coef_)

    # Lasso regression
    lasso_scores = []
    lasso_coefs = []

    for alpha in lambda_values:
        lasso = Lasso(alpha=alpha, max_iter=2000)
        scores = cross_val_score(lasso, X_train_scaled, y_train_scaled, cv=5, scoring='neg_mean_squared_error')
        lasso_scores.append(-scores.mean())
        
        lasso.fit(X_train_scaled, y_train_scaled)
        lasso_coefs.append(lasso.coef_)

    # Plot results
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))

    # Cross-validation scores
    axes[0, 0].semilogx(lambda_values, ridge_scores, 'b-', label='Ridge')
    axes[0, 0].semilogx(lambda_values, lasso_scores, 'r-', label='Lasso')
    axes[0, 0].set_xlabel('Regularization Parameter (λ)')
    axes[0, 0].set_ylabel('Cross-Validation MSE')
    axes[0, 0].set_title('Cross-Validation Performance')
    axes[0, 0].legend()
    axes[0, 0].grid(True)

    # Coefficient paths
    ridge_coefs = np.array(ridge_coefs)
    lasso_coefs = np.array(lasso_coefs)

    axes[0, 1].semilogx(lambda_values, ridge_coefs)
    axes[0, 1].set_xlabel('Regularization Parameter (λ)')
    axes[0, 1].set_ylabel('Coefficient Values')
    axes[0, 1].set_title('Ridge: Coefficient Paths')
    axes[0, 1].grid(True)

    axes[1, 0].semilogx(lambda_values, lasso_coefs)
    axes[1, 0].set_xlabel('Regularization Parameter (λ)')
    axes[1, 0].set_ylabel('Coefficient Values')
    axes[1, 0].set_title('Lasso: Coefficient Paths')
    axes[1, 0].grid(True)

    # Sparsity comparison
    ridge_nonzero = np.sum(ridge_coefs != 0, axis=1)
    lasso_nonzero = np.sum(lasso_coefs != 0, axis=1)

    axes[1, 1].semilogx(lambda_values, ridge_nonzero, 'b-', label='Ridge')
    axes[1, 1].semilogx(lambda_values, lasso_nonzero, 'r-', label='Lasso')
    axes[1, 1].set_xlabel('Regularization Parameter (λ)')
    axes[1, 1].set_ylabel('Number of Non-zero Coefficients')
    axes[1, 1].set_title('Sparsity Comparison')
    axes[1, 1].legend()
    axes[1, 1].grid(True)

    plt.tight_layout()
    plt.show()

    # Optimal lambda selection
    best_ridge_idx = np.argmin(ridge_scores)
    best_lasso_idx = np.argmin(lasso_scores)

    print(f"Best Ridge λ: {lambda_values[best_ridge_idx]:.4f}")
    print(f"Best Lasso λ: {lambda_values[best_lasso_idx]:.4f}")

    # Final model evaluation
    best_ridge = Ridge(alpha=lambda_values[best_ridge_idx])
    best_lasso = Lasso(alpha=lambda_values[best_lasso_idx])

    best_ridge.fit(X_train_scaled, y_train_scaled)
    best_lasso.fit(X_train_scaled, y_train_scaled)

    # Transform coefficients back to original scale
    ridge_coef_original = best_ridge.coef_ * scaler_y.scale_ / scaler_X.scale_
    lasso_coef_original = best_lasso.coef_ * scaler_y.scale_ / scaler_X.scale_

    ridge_intercept = scaler_y.mean_ - np.sum(ridge_coef_original * scaler_X.mean_)
    lasso_intercept = scaler_y.mean_ - np.sum(lasso_coef_original * scaler_X.mean_)

    print("\nRidge Regression Results:")
    print(f"Intercept: {ridge_intercept:.4f}")
    print(f"Non-zero coefficients: {np.sum(ridge_coef_original != 0)}")
    print(f"Test R²: {r2_score(y_test, ridge_intercept + X_test @ ridge_coef_original):.4f}")

    print("\nLasso Regression Results:")
    print(f"Intercept: {lasso_intercept:.4f}")
    print(f"Non-zero coefficients: {np.sum(lasso_coef_original != 0)}")
    print(f"Test R²: {r2_score(y_test, lasso_intercept + X_test @ lasso_coef_original):.4f}")
    
    return ridge_coef_original, lasso_coef_original, lambda_values, ridge_scores, lasso_scores

# Run demonstration
ridge_coef_original, lasso_coef_original, lambda_values, ridge_scores, lasso_scores = demonstrate_regularization_comparison()

File: 03_variable_selection_regularization/code/regularization_comparison.R
# Regularization Comparison: Ridge vs Lasso
# This script demonstrates ridge and lasso regression using the glmnet package

# Load required libraries
library(glmnet)
library(ggplot2)
library(dplyr)

# Set random seed for reproducibility
set.seed(42)

# Generate synthetic data
n <- 100
p <- 20
X <- matrix(rnorm(n * p), n, p)
true_beta <- rep(0, p)
true_beta[1:5] <- c(2, -1.5, 1, -0.8, 0.6)
y <- X %*% true_beta + 0.5 * rnorm(n)

# Split data into training and test sets
train_idx <- sample(1:n, 0.7 * n)
X_train <- X[train_idx, ]
X_test <- X[-train_idx, ]
y_train <- y[train_idx]
y_test <- y[-train_idx]

cat("=== REGULARIZATION COMPARISON: RIDGE VS LASSO ===\n")
cat("Sample size:", n, "\n")
cat("Number of predictors:", p, "\n")
cat("Training set size:", length(train_idx), "\n")
cat("Test set size:", n - length(train_idx), "\n")
cat("True non-zero coefficients:", sum(true_beta != 0), "\n\n")

# Fit ridge regression with cross-validation
cat("=== RIDGE REGRESSION ===\n")
ridge_cv <- cv.glmnet(X_train, y_train, alpha = 0, standardize = TRUE)
ridge_fit <- glmnet(X_train, y_train, alpha = 0, lambda = ridge_cv$lambda.min)

cat("Best ridge lambda:", round(ridge_cv$lambda.min, 4), "\n")
cat("Ridge CV MSE:", round(min(ridge_cv$cvm), 4), "\n")

# Fit lasso regression with cross-validation
cat("\n=== LASSO REGRESSION ===\n")
lasso_cv <- cv.glmnet(X_train, y_train, alpha = 1, standardize = TRUE)
lasso_fit <- glmnet(X_train, y_train, alpha = 1, lambda = lasso_cv$lambda.min)

cat("Best lasso lambda:", round(lasso_cv$lambda.min, 4), "\n")
cat("Lasso CV MSE:", round(min(lasso_cv$cvm), 4), "\n")

# Plot coefficient paths
par(mfrow = c(1, 2))

# Ridge coefficient paths
plot(ridge_cv$glmnet.fit, xvar = "lambda", main = "Ridge: Coefficient Paths")
abline(v = log(ridge_cv$lambda.min), col = "red", lty = 2)

# Lasso coefficient paths
plot(lasso_cv$glmnet.fit, xvar = "lambda", main = "Lasso: Coefficient Paths")
abline(v = log(lasso_cv$lambda.min), col = "red", lty = 2)

# Make predictions on test set
ridge_pred <- predict(ridge_fit, newx = X_test)
lasso_pred <- predict(lasso_fit, newx = X_test)

# Calculate R-squared for both models
ridge_r2 <- 1 - sum((y_test - ridge_pred)^2) / sum((y_test - mean(y_test))^2)
lasso_r2 <- 1 - sum((y_test - lasso_pred)^2) / sum((y_test - mean(y_test))^2)

# Calculate MSE for both models
ridge_mse <- mean((y_test - ridge_pred)^2)
lasso_mse <- mean((y_test - lasso_pred)^2)

# Model comparison results
cat("\n=== MODEL COMPARISON ===\n")
cat("Ridge R²:", round(ridge_r2, 4), "\n")
cat("Lasso R²:", round(lasso_r2, 4), "\n")
cat("Ridge MSE:", round(ridge_mse, 4), "\n")
cat("Lasso MSE:", round(lasso_mse, 4), "\n")
cat("Ridge non-zero coefficients:", sum(coef(ridge_fit) != 0), "\n")
cat("Lasso non-zero coefficients:", sum(coef(lasso_fit) != 0), "\n")

# Extract coefficients for comparison
ridge_coef <- as.vector(coef(ridge_fit))
lasso_coef <- as.vector(coef(lasso_fit))

# Create coefficient comparison plot
coef_comparison <- data.frame(
  Variable = 0:p,
  Ridge = ridge_coef,
  Lasso = lasso_coef,
  True = c(0, true_beta)  # Include intercept
)

# Plot coefficient comparison
ggplot(coef_comparison, aes(x = Variable)) +
  geom_line(aes(y = Ridge, color = "Ridge"), size = 1) +
  geom_line(aes(y = Lasso, color = "Lasso"), size = 1) +
  geom_point(aes(y = True, color = "True"), size = 2) +
  scale_color_manual(values = c("Ridge" = "blue", "Lasso" = "red", "True" = "green")) +
  labs(title = "Coefficient Comparison: Ridge vs Lasso vs True",
       x = "Variable Index",
       y = "Coefficient Value",
       color = "Model") +
  theme_minimal() +
  theme(legend.position = "bottom")

# Print coefficient summary
cat("\n=== COEFFICIENT SUMMARY ===\n")
cat("True non-zero coefficients (indices):", which(true_beta != 0), "\n")
cat("Ridge non-zero coefficients (indices):", which(ridge_coef[-1] != 0), "\n")
cat("Lasso non-zero coefficients (indices):", which(lasso_coef[-1] != 0), "\n")

# Calculate coefficient accuracy
ridge_accuracy <- sum((ridge_coef[-1] != 0) == (true_beta != 0)) / p
lasso_accuracy <- sum((lasso_coef[-1] != 0) == (true_beta != 0)) / p

cat("\n=== VARIABLE SELECTION ACCURACY ===\n")
cat("Ridge variable selection accuracy:", round(ridge_accuracy, 4), "\n")
cat("Lasso variable selection accuracy:", round(lasso_accuracy, 4), "\n")

# Key insights
cat("\n=== KEY INSIGHTS ===\n")
cat("1. Ridge regression keeps all variables but shrinks coefficients\n")
cat("2. Lasso regression performs automatic variable selection\n")
cat("3. Cross-validation helps select optimal regularization parameters\n")
cat("4. Lasso provides sparsity while ridge provides stability\n")
cat("5. Both methods can improve prediction accuracy over OLS\n")

# Return results for further analysis
results <- list(
  ridge_cv = ridge_cv,
  lasso_cv = lasso_cv,
  ridge_fit = ridge_fit,
  lasso_fit = lasso_fit,
  ridge_coef = ridge_coef,
  lasso_coef = lasso_coef,
  ridge_r2 = ridge_r2,
  lasso_r2 = lasso_r2,
  ridge_mse = ridge_mse,
  lasso_mse = lasso_mse,
  true_beta = true_beta
)

cat("\n=== ANALYSIS COMPLETE ===\n")
cat("Results stored in 'results' object for further analysis\n")
